nifi-nar-bundles/nifi-smb-bundle/nifi-smb-connection-pool/src/main/java/org/apache/nifi/services/smb/SmbjSessionProviderService.java

```diff
@@ -0,0 +1,175 @@
+            .sensitive(true)
+            .build();
+    public static final PropertyDescriptor PORT = new PropertyDescriptor.Builder()
+            .displayName("Port")
+            .name("port")
+            .description("Port to use for connection.")
+            .required(true)
+            .addValidator(PORT_VALIDATOR)
+            .defaultValue("445")
+            .build();
+    public static final PropertyDescriptor TIMEOUT = new PropertyDescriptor.Builder()
+            .displayName("Timeout")
+            .name("timeout")
+            .description("Timeout in seconds for read and write operations.")
+            .required(true)
+            .defaultValue("5")
+            .addValidator(INTEGER_VALIDATOR)
+            .build();
+    private static final List<PropertyDescriptor> PROPERTIES = Collections
+            .unmodifiableList(asList(
+                    HOSTNAME,
+                    DOMAIN,
+                    USERNAME,
+                    PASSWORD,
+                    PORT,
+                    TIMEOUT
+            ));
+    private SMBClient smbClient;
+    private AuthenticationContext authenticationContext;
+    private ConfigurationContext context;
+    private String hostname;
+    private Integer port;
+
+    @Override
+    public Session getSession() {
+        try {
+            return smbClient.connect(hostname, port).authenticate(authenticationContext);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public URI getServiceLocation() {
+        return URI.create(String.format("smb://%s:%d", hostname, port));
+    }
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) throws IOException {
+        this.context = context;
+        this.hostname = context.getProperty(HOSTNAME).getValue();
+        this.port = context.getProperty(PORT).asInteger();
+        this.smbClient = new SMBClient(SmbConfig.builder()
+                .withTimeout(context.getProperty(TIMEOUT).asLong(), SECONDS)
+                .withTransportLayerFactory(new AsyncDirectTcpTransportFactory<>())
+                .build());
+        createAuthenticationContext();
+    }
+
+    @OnDisabled
+    public void onDisabled() {
+        smbClient.close();
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final String hostname = validationContext.getProperty(HOSTNAME).getValue();
+        final String port = validationContext.getProperty(PORT).getValue();
+        final String timeout = validationContext.getProperty(TIMEOUT).getValue();
+
+        return asList(HOSTNAME.validate(hostname, validationContext),
+                PORT.validate(port, validationContext),
+                TIMEOUT.validate(timeout, validationContext)
+        );
+    }
```

Comment:
This method can be removed since the standard validators work for each property.

Reply:
By adding this I override this function from AbstractConfigurableComponent:

    protected Collection<ValidationResult> customValidate(final ValidationContext validationContext) {
        return Collections.emptySet();
    }

Are you sure it will properly validate during enabling without this function?

Reply:
Yes, the `customValidate()` method works in addition to the validators defined on each PropertyDescriptor, so it is not necessary to implement `customValidate()` when standard validators are defined. A custom validation implementation can be used to evaluate dependencies between properties, or other logic, but the standard validators work when a property can be evaluate on its own.